### ISSUE TITLE
Multinomial logistic regression

### ISSUE THREAD

**Issue by jerrycraft2** [gorgonia/gorgonia#232]
Hi,
Need some help here.
I do not know how to change the logistic regression example to Multinomial logistic regression.

Example like
Which blood type does a person have, given the results of various diagnostic tests?

The output is multiple.

Thank you

**Comment by chewxy** [comment]
Typically this is done using softmax. 

Concretely, you'd want to replace G.Sigmoid in the following ine with G.Softmax

https://github.com/gorgonia/gorgonia/blob/master/examples/logisticregression/main.go#L92

Might also want to change your cost function

**Comment by jerrycraft2** [comment]
Hi,
Sorry as I am quite new to gorgonia concept. I get an error as below and no idea how to fix it.
panic: Failed to infer shape. Op: - false: Shape mismatch: (26733) and (26733, 4)
goroutine 1 [running]:
gorgonia.org/gorgonia.Must(0x0, 0xe9aaa0, 0xc000176540, 0x0, 0x0, 0x0, 0x0)
	C:/Users/jerry/go/src/gorgonia.org/gorgonia/gorgonia.go:19 +0xa7
main.main()
	C:/Users/jerry/go/src/gorgonia.org/gorgonia/examples/ocr/main.go:96 +0x8c2

package main

import (
	"flag"
	"fmt"
	"log"
	"math/rand"
	"os"
	"runtime/pprof"
	"time"

	G "gorgonia.org/gorgonia"
	"gorgonia.org/tensor"
)

const (
	// N is the number of rows in our dataset
	N = 26733
	// feats is the number of features (x) in our dataset
	feats = 10
	outs  = 4
	// trainIter is the number of interations for which to train
	trainIter = 500
)

var cpuprofile = flag.String("cpuprofile", "", "write cpu profile to file")
var memprofile = flag.String("memprofile", "", "write mem profile to file")
var static = flag.Bool("static", false, "Use static test file")
var wT tensor.Tensor
var yT tensor.Tensor
var xT tensor.Tensor

// in this example, we will generate random float64 values
var Float = tensor.Float64

// init generates random values for x, w, and y for demo purposes
// Use the static flag to load your own data
func init() {
	xBacking := tensor.Random(Float, N*feats)
	wBacking := tensor.Random(Float, feats)
	var yBacking interface{}
	switch Float {
	case tensor.Float64:
		backing := make([]float64, N*outs)
		for i := range backing {
			backing[i] = float64(rand.Intn(2))
		}
		yBacking = backing
	case tensor.Float32:
		backing := make([]float32, N*outs)
		for i := range backing {
			backing[i] = float32(rand.Intn(2))
		}
		yBacking = backing
	}

	xT = tensor.New(tensor.WithBacking(xBacking), tensor.WithShape(N, feats))
	yT = tensor.New(tensor.WithBacking(yBacking), tensor.WithShape(N, outs))
	wT = tensor.New(tensor.WithBacking(wBacking), tensor.WithShape(feats))
}

func main() {
	flag.Parse()
	rand.Seed(1337)
	log.SetFlags(0)

	if *static {
		// Float = tensor.Float64 // because the loadStatck function only loads []float64
		// wBacking, xBacking, yBacking := loadStatic()
		// xT = tensor.New(tensor.WithBacking(xBacking), tensor.WithShape(N, feats))
		// yT = tensor.New(tensor.WithBacking(yBacking), tensor.WithShape(N, outs))
		// wT = tensor.New(tensor.WithBacking(wBacking), tensor.WithShape(feats))
	}

	// To start, we need to create a graph and construct all the nodes.
	// Everything from the input to the prediction needs to be a node.

	// We start by creating nodes for the training
	// create a new graph and add x, y, w, b, and one as Nodes
	g := G.NewGraph()
	x := G.NewMatrix(g, Float, G.WithName("x"), G.WithShape(N, feats))
	y := G.NewMatrix(g, Float, G.WithName("y"), G.WithShape(N, outs))

	w := G.NewVector(g, Float, G.WithName("w"), G.WithShape(feats))
	b := G.NewScalar(g, Float, G.WithName("bias"))
	// Add a constant node 1 to be used later in the loss function
	//	one := G.NewConstant(1.0)

	// Here we create the nodes that will do the prediction operations
	// create a node that has the operation: (x*w + b)
	xwmb := G.Must(G.Add(G.Must(G.Mul(x, w)), b))
	// create a node that has the operation: SoftMax(xwmb)
	prob := G.Must(G.SoftMax(xwmb))
	G.WithName("prob")(prob)

	se := G.Must(G.Square(G.Must(G.Sub(prob, y))))
	cost := G.Must(G.Mean(se))
	G.WithName("cost")(cost)

	// calculate gradient by backpropagation https://en.wikipedia.org/wiki/Backpropagation
	grads, err := G.Grad(cost, w, b)
	handleError(err)
	// "dcost/dw" == derivative of cost with respect to w
	G.WithName("dcost/dw")(grads[0])
	// "dcost/db" == derivative of cost with respect to b
	G.WithName("dcost/db")(grads[1])

	// create the nodes for calculating the gradient
	learnRate := G.NewConstant(0.1) // be careful not to set a learnRate too high
	gwlr := G.Must(G.Mul(learnRate, grads[0]))
	wUpd := G.Must(G.Sub(w, gwlr))
	gblr := G.Must(G.Mul(learnRate, grads[1]))
	bUpd := G.Must(G.Sub(b, gblr))

	// Run the following line to write to the gographviz file for debugging https://github.com/awalterschulze/gographviz
	// ioutil.WriteFile("fullGraph.dot", []byte(g.ToDot()), 0644)

	// Now that we have created all the notes, we should compile a training program.
	// We are essentially creating a list of instructions to get from our inputs {x, y}
	// to our outputs {wUpd, bUpd, readNode}. Note that we need to tell gorgonia that
	// readNode is one of our outputs so that we can access it.
	prog, locMap, err := G.CompileFunction(g, G.Nodes{x, y}, G.Nodes{wUpd, bUpd})
	handleError(err)
	fmt.Printf("%v", prog) // print the instructions
	// With our program, we initialize a new TapeMachine that will execute our program
	machine := G.NewTapeMachine(g, G.WithPrecompiled(prog, locMap))
	defer machine.Close()
	// Note that NewTapeMachine() will compile if WithPrecomiled() is not provided.
	// Internally, Gorgonia will figure out that a compilation process needs to happen,
	// so it will call Compile(g), which will output the prog and locMap internally.
	// When nodes are added to the graph to Gorgonia, nodes that have no precedents
	// and have a nil Op are marked as input nodes. So Gorgonia knows to start there.
	// But since our graph contains the training and prediction nodes, we manually compiled.

	// we allocated the node w before, but we never set it's initial value
	machine.Let(w, wT)
	machine.Let(b, 0.0)

	if *cpuprofile != "" {
		f, err := os.Create(*cpuprofile)
		if err != nil {
			log.Fatal(err)
		}
		pprof.StartCPUProfile(f)
		defer pprof.StopCPUProfile()
	}

	// Now that we have our graph, program, and machine, we can start training
	start := time.Now()
	for i := 0; i < trainIter; i++ {
		// move the pointer back to the beginning of the prog. Reset() does not delete any values
		machine.Reset()
		// We should reinitialize the values {x,y}. This is a good practice.
		// Think of the machine as a function(x,y) and we are providing the input values
		machine.Let(x, xT)
		machine.Let(y, yT)
		handleError(machine.RunAll())
		// After running the machine, we want to update w and b
		machine.Set(w, wUpd)
		machine.Set(b, bUpd)

		// After each iteration, we print out the training accuracy to see
		// how our algorithm is doing
	}
	fmt.Printf("Time taken: %v\n", time.Since(start))
	fmt.Printf("Final Model: \nw: %3.3s\nb: %+3.3s\n", w.Value(), b.Value())

}

func handleError(err error) {
	if err != nil {
		log.Fatalf("%+v", err)
	}
}

**Comment by jerrycraft2** [comment]
Hi,

Sorry as I am new in Neural Network. How do I create the Graph if I have 5 feature in the input.
And 4 output. Can you please help pointing where is my mistakes. Thank you.

Example : 
Which blood type (4 type) does a person have, given the results of various diagnostic tests (5 feature)?

	g := G.NewGraph()
	x := G.NewMatrix(g, Float, G.WithName("x"), G.WithShape(N, feats))
	y := G.NewMatrix(g, Float, G.WithName("y"), G.WithShape(N, outs))
	w := G.NewMatrix(g, Float, G.WithName("w"), G.WithShape(feats, outs))
	b := G.NewScalar(g, Float, G.WithName("bias"))
	// Add a constant node 1 to be used later in the loss function
	//	one := G.NewConstant(1.0)

	// Here we create the nodes that will do the prediction operations
	// create a node that has the operation: (x*w + b)
	xwmb := G.Must(G.Add(G.Must(G.Mul(x, w)), b))
	// create a node that has the operation: SoftMax(xwmb)
	prob := G.Must(G.SoftMax(xwmb))
	G.WithName("prob")(prob)

	se := G.Must(G.Square(G.Must(G.Sub(prob, y))))
	cost := G.Must(G.Mean(se))
	G.WithName("cost")(cost)

**Comment by chewxy** [comment]
Why is your bias a scalar

**Comment by jerrycraft2** [comment]
Hi,

Appreciate your help. But I still have mistake somewhere.
I have change bias to shape of (N, outs). My setup is N=10, feats=3 and outs=4
Final Model after running: 
w: ⎡   1.76   -0.701    0.795   -0.351⎤
⎢ -0.744     1.46     1.08   -0.287⎥
⎣-0.0694    0.609   -0.435   -0.105⎦
b: ⎡  0.447  -0.0704   -0.288  -0.0891⎤
⎢-0.0535    0.315   -0.194  -0.0676⎥
  ⋮  
⎢  0.447  -0.0704   -0.288  -0.0891⎥
⎣-0.0535    0.315   -0.194  -0.0676⎦

My final model, w shape is 3 x 4 and bias shape is 10 x 4
N=10 sample test is small right now. if my sample is 1000, then my bias shape will be 1000 x 4
Is this large bias shape correct?

I am confuse of which to use, NewTensor or NewMatrix for this Multinomial logistic regression.
For bias, I am using NewTensor to replace the NewScalar. I do not know if this is correct.
Thank you in advance for your help. 👍 

package main

import (
	"flag"
	"fmt"
	"log"
	"math/rand"
	"os"
	"runtime/pprof"
	"time"

	G "gorgonia.org/gorgonia"
	"gorgonia.org/tensor"
)

const (
	// N is the number of rows in our dataset
	N = 10
	// feats is the number of features (x) in our dataset
	feats = 3
	// outs is number of output (y) in our dataset
	outs = 4
	// trainIter is the number of interations for which to train
	trainIter = 1000
)

var cpuprofile = flag.String("cpuprofile", "", "write cpu profile to file")
var memprofile = flag.String("memprofile", "", "write mem profile to file")
var static = flag.Bool("static", true, "Use static test file")
var wT tensor.Tensor
var yT tensor.Tensor
var xT tensor.Tensor

// in this example, we will generate random float64 values
var Float = tensor.Float64

// init generates random values for x, w, and y for demo purposes
// Use the static flag to load your own data
func init() {
	xBacking := tensor.Random(Float, N*feats)
	wBacking := tensor.Random(Float, feats*outs)
	var yBacking interface{}
	switch Float {
	case tensor.Float64:
		backing := make([]float64, N*outs)
		for i := range backing {
			backing[i] = float64(rand.Intn(2))
		}
		yBacking = backing
	case tensor.Float32:
		backing := make([]float32, N*outs)
		for i := range backing {
			backing[i] = float32(rand.Intn(2))
		}
		yBacking = backing
	}

	xT = tensor.New(tensor.WithBacking(xBacking), tensor.WithShape(N, feats))
	yT = tensor.New(tensor.WithBacking(yBacking), tensor.WithShape(N, outs))
	wT = tensor.New(tensor.WithBacking(wBacking), tensor.WithShape(feats, outs))
}

func main() {
	flag.Parse()
	rand.Seed(1337)
	log.SetFlags(0)

	if *static {
		Float = tensor.Float64 // because the loadStatck function only loads []float64
		wBacking, xBacking, yBacking := loadStatic()
		xT = tensor.New(tensor.WithBacking(xBacking), tensor.WithShape(N, feats))
		yT = tensor.New(tensor.WithBacking(yBacking), tensor.WithShape(N, outs))
		wT = tensor.New(tensor.WithBacking(wBacking), tensor.WithShape(feats, outs))
	}

	// To start, we need to create a graph and construct all the nodes.
	// Everything from the input to the prediction needs to be a node.

	// We start by creating nodes for the training
	// create a new graph and add x, y, w, b, and one as Nodes
	g := G.NewGraph()
	x := G.NewMatrix(g, Float, G.WithName("x"), G.WithShape(N, feats))
	y := G.NewMatrix(g, Float, G.WithName("y"), G.WithShape(N, outs))
	w := G.NewMatrix(g, Float, G.WithName("w"), G.WithShape(feats, outs))
	b := G.NewTensor(g, Float, 2, G.WithName("bias"), G.WithShape(N, outs), G.WithInit(G.Zeroes()))
	// Add a constant node 1 to be used later in the loss function
	//	one := G.NewConstant(1.0)

	// Here we create the nodes that will do the prediction operations
	// create a node that has the operation: (x*w + b)
	xwmb := G.Must(G.Add(G.Must(G.Mul(x, w)), b))
	//	xwmb2 := G.Must(G.Mul(xwmb, w0))
	// create a node that has the operation: SoftMax(xwmb)
	prob := G.Must(G.SoftMax(xwmb))
	G.WithName("prob")(prob)

	se := G.Must(G.Square(G.Must(G.Sub(prob, y))))
	cost := G.Must(G.Mean(se))
	G.WithName("cost")(cost)

	// calculate gradient by backpropagation https://en.wikipedia.org/wiki/Backpropagation
	grads, err := G.Grad(cost, w, b)
	handleError(err)
	// "dcost/dw" == derivative of cost with respect to w
	G.WithName("dcost/dw")(grads[0])
	// "dcost/db" == derivative of cost with respect to b
	G.WithName("dcost/db")(grads[1])

	// create the nodes for calculating the gradient
	learnRate := G.NewConstant(0.1) // be careful not to set a learnRate too high
	gwlr := G.Must(G.Mul(learnRate, grads[0]))
	wUpd := G.Must(G.Sub(w, gwlr))
	gblr := G.Must(G.Mul(learnRate, grads[1]))
	bUpd := G.Must(G.Sub(b, gblr))

	// Run the following line to write to the gographviz file for debugging https://github.com/awalterschulze/gographviz
	// ioutil.WriteFile("fullGraph.dot", []byte(g.ToDot()), 0644)

	// Now that we have created all the notes, we should compile a training program.
	// We are essentially creating a list of instructions to get from our inputs {x, y}
	// to our outputs {wUpd, bUpd, readNode}. Note that we need to tell gorgonia that
	// readNode is one of our outputs so that we can access it.
	prog, locMap, err := G.CompileFunction(g, G.Nodes{x, y}, G.Nodes{wUpd, bUpd})
	handleError(err)
	fmt.Printf("%v", prog) // print the instructions
	// With our program, we initialize a new TapeMachine that will execute our program
	machine := G.NewTapeMachine(g, G.WithPrecompiled(prog, locMap))
	defer machine.Close()
	// Note that NewTapeMachine() will compile if WithPrecomiled() is not provided.
	// Internally, Gorgonia will figure out that a compilation process needs to happen,
	// so it will call Compile(g), which will output the prog and locMap internally.
	// When nodes are added to the graph to Gorgonia, nodes that have no precedents
	// and have a nil Op are marked as input nodes. So Gorgonia knows to start there.
	// But since our graph contains the training and prediction nodes, we manually compiled.

	// we allocated the node w before, but we never set it's initial value
	machine.Let(w, wT)
	//	machine.Let(b, 0.0)

	if *cpuprofile != "" {
		f, err := os.Create(*cpuprofile)
		if err != nil {
			log.Fatal(err)
		}
		pprof.StartCPUProfile(f)
		defer pprof.StopCPUProfile()
	}

	// Now that we have our graph, program, and machine, we can start training
	start := time.Now()
	for i := 0; i < trainIter; i++ {
		// move the pointer back to the beginning of the prog. Reset() does not delete any values
		machine.Reset()
		// We should reinitialize the values {x,y}. This is a good practice.
		// Think of the machine as a function(x,y) and we are providing the input values
		machine.Let(x, xT)
		machine.Let(y, yT)
		handleError(machine.RunAll())
		// After running the machine, we want to update w and b
		machine.Set(w, wUpd)
		machine.Set(b, bUpd)

		// After each iteration, we print out the training accuracy to see
		// how our algorithm is doing
		// fmt.Println(i, w.Value())
		// fmt.Println(prob.Value())
	}
	fmt.Printf("Time taken: %v\n", time.Since(start))
	fmt.Printf("Final Model: \nw: %3.3s\nb: %3.3s", w.Value(), b.Value())

}

func handleError(err error) {
	if err != nil {
		log.Fatalf("%+v", err)
	}
}

**Comment by chewxy** [comment]
Gimme a few hrs to look into this. Busy day today

**Comment by chewxy** [comment]
> I am confuse of which to use, NewTensor or NewMatrix for this Multinomial logistic regression.
For bias, I am using NewTensor to replace the NewScalar. I do not know if this is correct.
Thank you in advance for your help

There's no difference. `NewMatrix`, `NewVector` are helper functions. They are essentially the same as `NewTensor`

**Comment by chewxy** [comment]
regarding your question about the size of your bias, we're working on an alternative that allows you to do a `BroadcastAdd`.

**Comment by chewxy** [comment]
Otherwise, your code looks fine

**Comment by jerrycraft2** [comment]
Hi,
I have solve the problem by removing the bias. Works perfectly. 
Just realize I do not need the bias. Thank you very much for your help. 👍
